File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/template/TemplateAbstract.java
package jv.com.tencent.cain.design_pattern.template;

/**
 * @author : jiangyu
 * @date : 2019/11/3
 * @desc :模版设计模式的模版父类
 */
public abstract class TemplateAbstract {
    public void templateMethod() {
        primitiveOperation1();
        primitiveOperation2();
    }
    /**
     * 留给具体的子类去实现的方法
     */
    abstract void primitiveOperation1();

    abstract void primitiveOperation2();
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/template/TemplateAbstract.kt
package kt.com.tencent.cain.desigin_pattern.template

/**
 * @author : jiangyu
 * @date   : 2019/11/3
 * @desc   : 模版设计模式的模版父类
 */
abstract class TemplateAbstract {
    /**
     * 留给具体的子类去实现的方法
     */
    abstract fun primitiveOperation1()

    abstract fun primitiveOperation2()

    fun templateMethod() {
        primitiveOperation1()
        primitiveOperation2()
    }
}